### PR TITLE
feat: Compress Floats columns containg natural numbers

### PR DESCRIPTION
**Pull request by e-dard**
This PR contributes to overall query performance because it allows us to fit more data into memory through greater compression of float columns.

## Rationale

Time-series data often contains floating-point data, and in the field a significant proportion of this is actually integer data that is provided as 64-bit floats—see Prometheus.

For integer data the Read Buffer already does Byte Trimming, which effectively stores 64-bit integer data as smaller phsical types (`i8`, `i16` etc). This can reduce columnar footprint 8x in the best cases. However, it is not safe to "trim" double-precision floating point data to 32-bit single-precision data because the conversion could be lossy (due to less available precision).

However, if all of the float values in a column are actually natural numbers with a certain size and range then they can be safely represented as 32-bit or smaller integers. Once you can do that then you can also do integer byte trimming and still RLE the result.

For example: you could go from something like this:

```
100.0 x1000 times, 101.0 x2000 times, 104.0 x1000 times  // a 4,000 element column of `~32KB`
```

to something like this:

```
// convert floats to integers then byte trim to `u8` then RLE

(1000, 100)
(2000, 101)
(1000, 104)
```

Which would end up taking up `~18b` in memory.

## Approach

The approach is pretty straightforward, building on the work done in #1563 and #1596. Effectively we implement a new `Transcoder` (`FloatByteTrimmer`), which encodes `f64` values as smaller integer representations, and then deocodes them back.

Then we teach the `FloatEncoder` to analyse input data and determine if it's suitable to be converted to integers and/or RLE/byte trimmed.

### CODE REVIEW

File: read_buffer/src/column/encoding/scalar/transcoders.rs [hunk 1]
```diff
@@ -74,12 +74,46 @@ impl Display for ByteTrimmer {
     }
 }
 
+/// An encoding that forcefully converts logical `f64` values into other integer
+/// types. It is the caller's responsibility to ensure that conversion can take
+/// place without loss of precision.
+#[derive(Debug)]
+pub struct FloatByteTrimmer {}
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/transcoders.rs:81]: A transcoder for converting floats into integers. The assertion shouldn't impact performance.

**Review by domodwyer — APPROVED** [review]

**Comment by crepererum** [comment]
Why does byte trimming not work on floats?

**Comment by e-dard** [comment]
@crepererum and I discussed the future extension of transmuted other float values into integers.

File: read_buffer/src/column/float.rs [hunk 2]
```diff
@@ -208,105 +215,468 @@ impl std::fmt::Display for FloatEncoding {
...
-        Self::F64(enc)
+        let name = enc.name();
+        Self::F64(enc, name.to_owned())
     }
 }
 
+// Convert float data to byte-trimmed integers and potentially RLE. It is the
+// caller's responsibility to ensure all data are natural numbers that can be
+// round tripped from float to int and back without loss of precision.
+fn from_array_with_byte_trimming(arr: Float64Array, range: (f64, f64)) -> FloatEncoding {
```
Comment (alamb) [inline read_buffer/src/column/float.rs:476]: it is a shame we need to duplicate the match logic here, though I can see how trying to make it generic might be challenging (as you need really need a SizedIterator over floats)

**Review by alamb — COMMENTED** [review]
This is pretty cool stuff. 👍  Nice work

**Pull request merged**